py-polars/polars/internals/construction.py

```diff
@@ -1229,15 +1229,40 @@ def to_frame_chunk(
     return (df.rechunk() if n_chunks > 0 else df)._df
 
 
+def pandas_has_default_index(df: pd.DataFrame) -> bool:
+    """Identify if the pandas frame only has a default (or equivalent) index."""
+    from pandas.core.indexes.numeric import IntegerIndex
+    from pandas.core.indexes.range import RangeIndex
+
+    index_cols = df.index.names
+
+    if len(index_cols) > 1:
+        return False  # not default: more than one index
+    elif index_cols not in ([None], [""]):
+        return False  # not default: index is named
+    elif df.index.equals(RangeIndex(start=0, stop=len(df), step=1)):
+        return True  # is default: simple range index
+    else:
+        # finally, is the index _equivalent_ to a default unnamed
+        # integer index with frame data that was previously sorted
+        return isinstance(df.index, IntegerIndex) and sorted(
+            df.index.values.tolist()
+        ) == list(range(len(df)))
+
+
 @deprecated_alias(nan_to_none="nan_to_null")
 def pandas_to_pydf(
     data: pd.DataFrame,
     schema: SchemaDefinition | None = None,
     schema_overrides: SchemaDict | None = None,
     rechunk: bool = True,
     nan_to_null: bool = True,
+    include_index: bool = False,
 ) -> PyDataFrame:
     """Construct a PyDataFrame from a pandas DataFrame."""
+    if include_index and not pandas_has_default_index(data):
+        data = data.reset_index(inplace=False)
```

Comment:
Preferably don't use `reset_index` as it will make a copy of the dataframe.
Convert the dataframe like normal and add the index as extra columns later.

Reply:
```
    if include_index and not pandas_has_default_index(data):
        index_df = pl.DataFrame(data.index.to_frame())
```

And later:
```
index_df.hstack(pl.DataFrame(pd_df))
```

Reply:
Ouch; for some reason I could have sworn `reset_index` didn't copy the underlying data... :(
Should be able to employ some minor shenanigans and do it with only inplace ops, hmm.

Reply:
Maybe it doesn't copy the data, but I seem to remember that I had this ones when loading data from a feather file with pandas and running `set_inplace(inplace=True)` avoided a copy of a 180G dataframe.

Reply:
> Maybe it doesn't copy the data, but I seem to remember that I had this ones when loading data from a feather file with pandas and running `set_inplace(inplace=True)` avoided a copy of a 180G dataframe.

Explicitly making use of known `inplace` ops seems like the way to go 👍

Reply:
@alexander-beedie and @ghuls this is the current pandas default behavior (to copy the underlying the data) but it can be disabled by turning on the new [copy-on-write](https://pandas.pydata.org/docs/user_guide/options.html) feature via either one of these commands:

```python
pd.set_option("mode.copy_on_write", True)
pd.options.mode.copy_on_write = True
```

I don't see a downside to enabling this mode upon import but I haven't researched it thoroughly.

See [here](https://towardsdatascience.com/a-solution-for-inconsistencies-in-indexing-operations-in-pandas-b76e10719744) for a bit more explanation. Here's a quick test that I did to make sure that with `reset_index()` it does indeed work:

```python
import pandas as pd
import numpy as np
import psutil

#pd.set_option("mode.copy_on_write", True)

# get existing memory usage
_, _, _, offset, _ = psutil.virtual_memory()

df = pd.DataFrame({
    'a': np.zeros(int(1e8)),
    'b': np.zeros(int(1e8)),
    'c': np.zeros(int(1e8)),
}).set_index('a')

_, _, _, bytes1, _ = psutil.virtual_memory()
used1 = (bytes1-offset)/1e9 # increase after df1 creation

df2 = df.reset_index()
_, _, _, bytes2, _ = psutil.virtual_memory()
used2 = (bytes2-offset)/1e9 # increase after df1 & df2 creation

total3, available3, percent3, used3, free3 = psutil.virtual_memory()

print(f"Before reset_index(): {used1:0.1f}GB")
print(f"After reset_index(): {used2:0.1f}GB")
```
With the option disabled as above, I get:
```
Before reset_index(): 2.4GB
After reset_index(): 4.8GB
```
When enabled:
```
Before reset_index(): 2.4GB
After reset_index(): 3.2GB
```
So the increase in memory is about 50% what it was. Something is obviously still getting copied, but it's not as bad. So...maybe @ghuls' workaround is still the way to go.

Reply:
> Explicitly making use of known inplace ops seems like the way to go 👍

This is discouraged by pandas unfortunately, and the `inplace` parameter [is going to be deprecated across the board in pandas 2.0](https://github.com/pandas-dev/pandas/issues/16529), but I think we should still do it. IMO it's silly to push for a deprecation of something prior to having a performant solution in place.

Reply:
Finally! @ghuls / @mcrumiller, found a (mostly) zero-copy way to get a hold the index data that doesn't rely on pandas >= 1.5x; adapted the script above to quickly check the before/after memory delta of this method, and it was `0.0GB` across the board. Integrating now 👍 
```
idxdata = df.index.get_level_values(idxcol)
```

Reply:
Awesome! I knew about `get_level_values()` (We use a LOT of MultiIndex here) but never realized it was 0-copy, that's actually great to know.

Reply:
Yup; looping that call and appending the results into a list 1,000x on your example frame above was essentially instant, with no memory delta:
```
Elapsed time: 0.0002 seconds
Delta: 0.0 GB

sum(len(i) for i in idxs)
100000000000
```